Possible regression in 3.1: destructured parameter is not recognized if it comes from intersection type

**TypeScript Version:**  3.2.0-dev.20181011

**Search Terms:** destructuring inference parameters intersection

**Code**

```ts

type MethodBase = { p: {}, needsPrep: 'yes' | 'no', prepArgs: {} };
type Methods = { [m in string]: MethodBase };

interface OptionsForMethodTypes<M extends MethodBase> {
    'yes': { prep: (args: { p: M['p'] } & M['prepArgs']) => void };
    'no': { };
}
type OptionsForMethod<M extends MethodBase> = OptionsForMethodTypes<M>[M['needsPrep']];

type Callers<MM extends Methods> = { [m in keyof MM]: (p: MM[m]['p']) => void };

type Options<MM extends Methods> = { [m in keyof MM]?: OptionsForMethod<MM[m]> };

declare function createCallers<MM extends Methods>(methods: MM, options?: Options<MM>): Callers<MM>;

const someMethods = {
    a: {} as { p: string, needsPrep: 'yes', prepArgs: {foo: number} },
    b: {} as { p: number, needsPrep: 'no', prepArgs: {} }
}

const c = createCallers(someMethods, {
    a: { prep: ({ p, foo }) => { } },

  // ERROR:  Type '{ p: unknown; }' has no property 'foo' and no string index signature.
  // at the same time, correct type for foo is shown in the same tooltip:
  // var foo: number
});


```

**Expected behavior:**

No error. This code compiles without errors in 3.0.1

**Actual behavior:**

ERROR:  Type '{ p: unknown; }' has no property 'foo' and no string index signature.

**Playground Link:**   [link](https://www.typescriptlang.org/play/#src=%0D%0Atype%20MethodBase%20%3D%20%7B%20p%3A%20%7B%7D%2C%20needsPrep%3A%20'yes'%20%7C%20'no'%2C%20prepArgs%3A%20%7B%7D%20%7D%3B%0D%0Atype%20Methods%20%3D%20%7B%20%5Bm%20in%20string%5D%3A%20MethodBase%20%7D%3B%0D%0A%0D%0Ainterface%20OptionsForMethodTypes%3CM%20extends%20MethodBase%3E%20%7B%0D%0A%20%20%20%20'yes'%3A%20%7B%20prep%3A%20(args%3A%20%7B%20p%3A%20M%5B'p'%5D%20%7D%20%26%20M%5B'prepArgs'%5D)%20%3D%3E%20void%20%7D%3B%0D%0A%20%20%20%20'no'%3A%20%7B%20%7D%3B%0D%0A%7D%0D%0Atype%20OptionsForMethod%3CM%20extends%20MethodBase%3E%20%3D%20OptionsForMethodTypes%3CM%3E%5BM%5B'needsPrep'%5D%5D%3B%0D%0A%0D%0Atype%20Callers%3CMM%20extends%20Methods%3E%20%3D%20%7B%20%5Bm%20in%20keyof%20MM%5D%3A%20(p%3A%20MM%5Bm%5D%5B'p'%5D)%20%3D%3E%20void%20%7D%3B%0D%0A%0D%0Atype%20Options%3CMM%20extends%20Methods%3E%20%3D%20%7B%20%5Bm%20in%20keyof%20MM%5D%3F%3A%20OptionsForMethod%3CMM%5Bm%5D%3E%20%7D%3B%0D%0A%0D%0Adeclare%20function%20createCallers%3CMM%20extends%20Methods%3E(methods%3A%20MM%2C%20options%3F%3A%20Options%3CMM%3E)%3A%20Callers%3CMM%3E%3B%0D%0A%0D%0Aconst%20someMethods%20%3D%20%7B%0D%0A%20%20%20%20a%3A%20%7B%7D%20as%20%7B%20p%3A%20string%2C%20needsPrep%3A%20'yes'%2C%20prepArgs%3A%20%7Bfoo%3A%20number%7D%20%7D%2C%0D%0A%20%20%20%20b%3A%20%7B%7D%20as%20%7B%20p%3A%20number%2C%20needsPrep%3A%20'no'%2C%20prepArgs%3A%20%7B%7D%20%7D%0D%0A%7D%0D%0A%0D%0Aconst%20c%20%3D%20createCallers(someMethods%2C%20%7B%0D%0A%20%20%20%20a%3A%20%7B%20prep%3A%20(%7B%20p%2C%20foo%20%7D)%20%3D%3E%20%7B%20%7D%20%7D%2C%0D%0A%7D)%3B%0D%0A%0D%0A)

**Related Issues:** 

did not find anything

Here's a simplified repro that I think is the same issue:
```ts
type MethodNameToArgs = { [m: string]: {} };

function f<M extends MethodNameToArgs>(o: { [K in keyof M]: (args: M[K]) => void }): void {}

interface MyMethods extends MethodNameToArgs {
    a: { x: number };
}

f<MyMethods>({
    a(args) {},
});
```

`args` is correctly of type `{ x: number }`, but there's a compile error under `--strictFunctionTypes`:
```
src/a.ts:9:14 - error TS2345: Argument of type '{ a(args: { x: number; }): void; }' is not assignable to parameter of type '{ [x: string]: (args: {}) => void; a: (args: { x: number; }) => void; }'.
  Property 'a' is incompatible with index signature.
    Type '(args: { x: number; }) => void' is not assignable to type '(args: {}) => void'.
      Types of parameters 'args' and 'args' are incompatible.
        Type '{}' is not assignable to type '{ x: number; }'.
          Property 'x' is missing in type '{}'.

 9 f<MyMethods>({
                ~
10     a(args) {},
   ~~~~~~~~~~~~~~~
11 });
```

Seeing something in a React project that's potentially related to this. Simplified example:

```ts
// src/client/TypenameSwitcher.tsx
import * as React from 'react';

interface SelectableInputProps {
  __typename: 'SelectableInput';
  name: string;
  options: {
    name: string;
    value: string;
  }[];
}

const SelectableInput = ({ name, options }: SelectableInputProps) => (
  <select name={name}>
    {options.map(({ name, value }) => (
      <option key={value} value={value}>
        {name}
      </option>
    ))}
  </select>
);

interface TextInputProps {
  __typename: 'TextInput';
  name: string;
}

const TextInput = ({ name }: TextInputProps) => (
  <input name={name} type="text" />
);

type TypenameSwitcherProps = SelectableInputProps | TextInputProps;

const TypenameSwitcher = (props: TypenameSwitcherProps) => {
  const { __typename } = props;

  if (__typename === 'SelectableInput') {
    return <SelectableInput {...props} />;
  } else if (__typename === 'TextInput') {
    return <TextInput {...props} />;
  }

  return null;
};

export default TypenameSwitcher;

```

The `SelectableInput` and `TextInput` uses show some errors:

```shell
$ node_modules/.bin/tsc --version
Version 3.1.6
$ node_modules/.bin/tsc
src/client/TypenameSwitcher.tsx:37:13 - error TS2322: Type '{ __typename: "SelectableInput"; name: string; options: { name: string; value: string; }[]; } | { __typename: "TextInput"; name: string; }' is not assignable to type 'IntrinsicAttributes & SelectableInputProps'.
  Type '{ __typename: "TextInput"; name: string; }' is not assignable to type 'SelectableInputProps'.
    Property 'options' is missing in type '{ __typename: "TextInput"; name: string; }'.

37     return <SelectableInput {...props} />;
               ~~~~~~~~~~~~~~~

src/client/TypenameSwitcher.tsx:39:13 - error TS2322: Type '{ __typename: "SelectableInput"; name: string; options: { name: string; value: string; }[]; } | { __typename: "TextInput"; name: string; }' is not assignable to type 'IntrinsicAttributes & TextInputProps'.
  Type '{ __typename: "SelectableInput"; name: string; options: { name: string; value: string; }[]; }' is not assignable to type 'TextInputProps'.
    Types of property '__typename' are incompatible.
      Type '"SelectableInput"' is not assignable to type '"TextInput"'.

39     return <TextInput {...props} />;
               ~~~~~~~~~

```

Strangely enough, this totally builds without errors:

```diff
 const TypenameSwitcher = (props: TypenameSwitcherProps) => {
-  const { __typename } = props;
-
-  if (__typename === 'SelectableInput') {
+  if (props.__typename === 'SelectableInput') {
     return <SelectableInput {...props} />;
-  } else if (__typename === 'TextInput') {
+  } else if (props.__typename === 'TextInput') {
     return <TextInput {...props} />;
   }
 
   return null;
 };
```

Does this seem related to the union type/destructuring errors above? Or, should I open a new issue?

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -5425,7 +5425,16 @@ namespace ts {
 
         function getTypeOfVariableOrParameterOrProperty(symbol: Symbol): Type {
             const links = getSymbolLinks(symbol);
-            return links.type || (links.type = getTypeOfVariableOrParameterOrPropertyWorker(symbol));
+            if (!links.type) {
+                const type = getTypeOfVariableOrParameterOrPropertyWorker(symbol);
+                // For a contextually typed parameter it is possible that a type has already
+                // been assigned (in assignTypeToParameterAndFixTypeParameters), and we want
+                // to preserve this type.
+                if (!links.type) {
+                    links.type = type;
+                }
+            }
+            return links.type;
         }
 
         function getTypeOfVariableOrParameterOrPropertyWorker(symbol: Symbol) {
@@ -5469,7 +5478,7 @@ namespace ts {
                 if (symbol.flags & SymbolFlags.ValueModule) {
                     return getTypeOfFuncClassEnumModule(symbol);
                 }
-                return errorType;
+                return reportCircularityError(symbol);
             }
             let type: Type | undefined;
             if (isInJSFile(declaration) &&
@@ -5528,7 +5537,7 @@ namespace ts {
                 if (symbol.flags & SymbolFlags.ValueModule) {
                     return getTypeOfFuncClassEnumModule(symbol);
                 }
-                type = reportCircularityError(symbol);
+                return reportCircularityError(symbol);
             }
             return type;
         }

--- a/tests/baselines/reference/jsFileClassSelfReferencedProperty.types
+++ b/tests/baselines/reference/jsFileClassSelfReferencedProperty.types
@@ -4,11 +4,11 @@ export class StackOverflowTest {
 
   constructor () {
     this.testStackOverflow = this.testStackOverflow.bind(this)
->this.testStackOverflow = this.testStackOverflow.bind(this) : error
+>this.testStackOverflow = this.testStackOverflow.bind(this) : any
 >this.testStackOverflow : any
 >this : this
 >testStackOverflow : any
->this.testStackOverflow.bind(this) : error
+>this.testStackOverflow.bind(this) : any
 >this.testStackOverflow.bind : any
 >this.testStackOverflow : any
 >this : this

--- a/tests/baselines/reference/parserES5ForOfStatement18.types
+++ b/tests/baselines/reference/parserES5ForOfStatement18.types
@@ -1,5 +1,5 @@
 === tests/cases/conformance/parser/ecmascript5/Statements/parserES5ForOfStatement18.ts ===
 for (var of of of) { }
 >of : any
->of : error
+>of : any
 

--- a/tests/baselines/reference/parserES5ForOfStatement19.types
+++ b/tests/baselines/reference/parserES5ForOfStatement19.types
@@ -1,5 +1,5 @@
 === tests/cases/conformance/parser/ecmascript5/Statements/parserES5ForOfStatement19.ts ===
 for (var of in of) { }
 >of : any
->of : error
+>of : any
 

--- a/tests/baselines/reference/parserForOfStatement18.types
+++ b/tests/baselines/reference/parserForOfStatement18.types
@@ -1,5 +1,5 @@
 === tests/cases/conformance/parser/ecmascript6/Iterators/parserForOfStatement18.ts ===
 for (var of of of) { }
 >of : any
->of : error
+>of : any
 

--- a/tests/baselines/reference/parserForOfStatement19.types
+++ b/tests/baselines/reference/parserForOfStatement19.types
@@ -1,5 +1,5 @@
 === tests/cases/conformance/parser/ecmascript6/Iterators/parserForOfStatement19.ts ===
 for (var of in of) { }
 >of : any
->of : error
+>of : any
 

--- a/tests/baselines/reference/recur1.types
+++ b/tests/baselines/reference/recur1.types
@@ -15,7 +15,7 @@ salt.pepper = function() {}
 
 var cobalt = new cobalt.pitch();   
 >cobalt : any
->new cobalt.pitch() : error
+>new cobalt.pitch() : any
 >cobalt.pitch : any
 >cobalt : any
 >pitch : any

--- a/tests/baselines/reference/recursiveExportAssignmentAndFindAliasedType7.types
+++ b/tests/baselines/reference/recursiveExportAssignmentAndFindAliasedType7.types
@@ -14,7 +14,7 @@ import self = require("recursiveExportAssignmentAndFindAliasedType7_moduleD");
 
 var selfVar = self;
 >selfVar : any
->self : error
+>self : any
 
 export = selfVar;
 >selfVar : any

